hw/ip/otbn/dv/rig/rig/gens/edge_load_store.py

```diff
@@ -0,0 +1,235 @@
+                isinstance(self.bnsid.operands[2].op_type, ImmOperandType) and
+                self.bnsid.operands[2].op_type.signed and
+                isinstance(self.bnsid.operands[3].op_type, OptionOperandType) and
+                isinstance(self.bnsid.operands[4].op_type, OptionOperandType)):
+            raise RuntimeError('BN.LID instruction from instructions file is '
+                               'not the shape expected by the BadLoadStore'
+                               'generator.')
+
+        for insn in [self.sw, self.bnsid]:
+            weight = cfg.insn_weights.get(insn.mnemonic)
+            if weight > 0:
+                self.weights.append(weight)
+                self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 1:
+            return None
+
+        weights = self.weights
+        prog_insn = None
+        idx = random.choices(range(len(self.insns)), weights=weights)[0]
+
+        prog_insn = self.fill_insn(self.insns[idx], model)
+        if prog_insn is None:
+            return None
+
+        snippet = ProgSnippet(model.pc, [prog_insn])
+        snippet.insert_into_program(program)
+        model.update_for_insn(prog_insn)
+        model.pc += 4
+        return (snippet, False, model)
+
+    def fill_insn(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+        '''Try to pick one of BN.SID or SW instructions
+
+        '''
+
+        # Special-case BN store instruction by mnemonic. These use
+        # complicated indirect addressing, so it's probably more sensible to
+        # give them special code.
+        if insn.mnemonic in ['bn.sid']:
+            return self._fill_bn_sid(insn, model)
+        if insn.mnemonic in ['sw']:
+            return self._fill_sw(insn, model)
+
+        return None
+
+    def _fill_sw(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+
+        imm_op_type = self.sw.operands[1].op_type
+
+        # Determine the offset range for SW
+        offset_rng = imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for SW (-2048, 2047)
+        min_offset, max_offset = offset_rng
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        base = []
+
+        for reg_idx, reg_val in known_regs:
+            if (model.dmem_size + min_offset < reg_val and
+               reg_val <= model.dmem_size - min_offset + 4):
+                base.append((reg_idx, reg_val))
```

Comment:
If I understand this right, we're trying to hit the address `model.dmem_size - 4` and we're going to have something like `reg_val + immediate = model.dmem_size - 4`. Rearranging, I think this means that `model.dmem_size - 4 - reg_val ∈ [min_offset, max_offset]`. I think the test above needs a "min" turned into a "max" and I'd consider writing it in the form `a <= b <= c` to avoid having to repeat the "b" part.

Reply:
In that case just writing 
```
reg_val = u_reg_val - (1 << 32) if u_reg_val >> 31 else u_reg_val
if (model.dmem_size - reg_val - 4 in range(min_offset, max_offset)):
    base.append((reg_idx, reg_val))
```
would be enough, right?

Reply:
Yep, I agree! One minor nit: `range(a, b)` represents `[a, b-1]`, so you'd need `max_offset + 1`. (And you can drop the outer parens).